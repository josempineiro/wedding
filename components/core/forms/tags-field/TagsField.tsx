import { Tags } from "@/components/core/forms/tags-field/Tags";
import { Field, type FieldProps } from "@/components/core/forms/field/Field";
import { Input, type InputProps } from "@/components/core/forms/input/Input";

export interface TagsFieldProps
  extends Omit<InputProps<string>, "onChange" | "value"> {
  label?: FieldProps["label"];
  type?: "text" | "email" | "password";
  error?: FieldProps["error"];
  inputRef?: React.Ref<HTMLInputElement>;
  value: string[];
  onChange: (value: string[]) => void;
  onAdd: (value: string) => void;
  name: FieldProps["name"];
  maxTagsToShow?: number;
}

export function TagsField({
  label,
  error,
  name,
  inputRef,
  onChange,
  onAdd,
  disabled,
  readOnly,
  value,
  maxTagsToShow,
  ...rest
}: TagsFieldProps) {
  const handleAdd = (tag: string) => {
    onAdd(tag);
    onChange([...value, tag]);
  };

  const handleDelete = (deletedTag: string) => {
    onChange(value.filter((tag) => tag !== deletedTag));
  };
  const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (
      event.key === "Enter" &&
      (event.target as HTMLInputElement).value !== ""
    ) {
      handleAdd(event.currentTarget.value);
    }
  };
  return (
    <Field error={error} label={label} name={name}>
      {!disabled && !readOnly && (
        <Input<string>
          ref={inputRef}
          name={name}
          className="bg-white px-2 py-1"
          onKeyDown={handleKeyDown}
          {...rest}
        />
      )}
      <Tags
        tags={value}
        onClose={handleDelete}
        closable={!disabled && !readOnly}
        maxTagsToShow={maxTagsToShow}
        className="bg-white px-2 py-1"
      />
    </Field>
  );
}
