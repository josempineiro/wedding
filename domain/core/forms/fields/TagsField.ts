import { Field } from "./Field";

export interface TagsField extends Field<Array<string>> {
  type: "tags";
  placeholder?: string;
}
