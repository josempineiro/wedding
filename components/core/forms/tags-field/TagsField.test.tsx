import { render, fireEvent, getAllByRole } from "@testing-library/react";
import { TagsField } from "./TagsField";

describe("TagsField", () => {
  const onChangeMock = jest.fn();
  const onAddMock = jest.fn();

  const props = {
    name: "tags",
    label: "Tags",
    error: "Invalid tags",
    value: ["tag1", "tag2"],
    onChange: onChangeMock,
    onAdd: onAddMock,
  };

  beforeEach(() => {
    onChangeMock.mockClear();
    onAddMock.mockClear();
  });

  it("should render correctly", () => {
    const { container, getByRole } = render(<TagsField {...props} />);
    const input = getByRole("textbox") as HTMLInputElement;
    expect(container).toMatchSnapshot();
    expect(input).toBeInTheDocument();
  });

  it("should delete a tag when the close button is clicked", () => {
    const { getByLabelText, getByText, getByRole, getAllByRole } = render(
      <TagsField {...props} />
    );
    const deleteButton = getAllByRole("delete-tag");
    fireEvent.click(deleteButton[0]);
    expect(onChangeMock).toHaveBeenCalledWith(["tag2"]);
  });

  it("should add a new tag when Enter is pressed", () => {
    const { getByRole } = render(<TagsField {...props} />);
    const input = getByRole("textbox") as HTMLInputElement;
    fireEvent.change(input, { target: { value: "tag3" } });
    fireEvent.keyDown(input, { key: "Enter" });
    expect(onChangeMock).toHaveBeenCalledWith(["tag1", "tag2", "tag3"]);
    expect(onAddMock).toHaveBeenCalledWith("tag3");
  });

  it("should not add a new tag when Enter is pressed and input value is empty", () => {
    const { getByRole } = render(<TagsField {...props} />);
    const input = getByRole("textbox") as HTMLInputElement;
    fireEvent.change(input, { target: { value: "" } });
    fireEvent.keyDown(input, { key: "Enter" });
    expect(onChangeMock).not.toHaveBeenCalled();
    expect(onAddMock).not.toHaveBeenCalled();
  });

  it("should not delete a tag when disabled", () => {
    const { queryByRole } = render(<TagsField {...props} disabled />);
    const deleteButton = queryByRole("delete-tag");
    expect(deleteButton).toBeNull();
  });

  it("should not delete a tag when readOnly", () => {
    const { queryByRole } = render(<TagsField {...props} readOnly />);
    const deleteButton = queryByRole("delete-tag");
    expect(deleteButton).toBeNull();
  });
  it("should show all tags when the show more button is clicked", () => {
    const { getByText, getByRole } = render(
      <TagsField
        {...props}
        value={["tag1", "tag2", "tag3", "tag4"]}
        maxTagsToShow={2}
      />
    );
    const showMoreButton = getByRole("more-tags");
    fireEvent.click(showMoreButton);
    expect(getByRole("textbox")).toBeInTheDocument();
  });
});
