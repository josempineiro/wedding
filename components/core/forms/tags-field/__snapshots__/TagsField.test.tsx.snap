// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsField should render correctly 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <label
      aria-labelledby="tags"
      class="text-sm text-gray-700 px-2"
      for="tags"
    >
      Tags
    </label>
    <input
      class="bg-white px-2 py-1 text-md ring-none min-w-0 w-auto bg-transparent outline-none"
      name="tags"
      role="textbox"
    />
    <div
      class="bg-white px-2 py-1 flex flex-wrap gap-2 items-center"
    >
      <div
        class="inline-flex items-center rounded-sm gap-4 px-2 py-1"
        role="tag"
      >
        <span>
          tag1
        </span>
        <button
          class="flex flex-row gap-4 justify-center items-center transition-colors duration-200 ease-in-out px-2 py-1 text-sm py-2 h-8 w-8 rounded-full "
          role="delete-tag"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-minus "
            data-icon="minus"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432 256c0 17.7-14.3 32-32 32L48 288c-17.7 0-32-14.3-32-32s14.3-32 32-32l352 0c17.7 0 32 14.3 32 32z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
      <div
        class="inline-flex items-center rounded-sm gap-4 px-2 py-1"
        role="tag"
      >
        <span>
          tag2
        </span>
        <button
          class="flex flex-row gap-4 justify-center items-center transition-colors duration-200 ease-in-out px-2 py-1 text-sm py-2 h-8 w-8 rounded-full "
          role="delete-tag"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-minus "
            data-icon="minus"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432 256c0 17.7-14.3 32-32 32L48 288c-17.7 0-32-14.3-32-32s14.3-32 32-32l352 0c17.7 0 32 14.3 32 32z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
    <p>
      Invalid tags
    </p>
  </div>
</div>
`;
